catalyst-gateway/bin/src/metrics/memory.rs

```diff
@@ -1 +1,167 @@
 //! Metrics related to memory analytics.
+
+use std::{
+    alloc::System,
+    sync::atomic::{AtomicBool, Ordering},
+    thread,
+};
+
+use memory_stats::{memory_stats, MemoryStats};
+use stats_alloc::{Region, StatsAlloc, INSTRUMENTED_SYSTEM};
+
+use crate::settings::Settings;
+
+/// Use the instrumented allocator for gathering allocation statistics.
+/// Note: This wraps the global allocator.
+/// All structs that use the global allocator can be tracked.
+#[global_allocator]
+static GLOBAL: &StatsAlloc<System> = &INSTRUMENTED_SYSTEM;
+
+/// This is to prevent the init function from accidentally being called multiple times.
+static IS_INITIALIZED: AtomicBool = AtomicBool::new(false);
+
+/// Starts a background thread to periodically update memory metrics.
+///
+/// This function spawns a thread that updates the global `MemoryMetrics`
+/// structure at regular intervals defined by `UPDATE_INTERVAL_MILLI`.
+pub(crate) fn init_metrics_updater() {
+    if IS_INITIALIZED.swap(true, Ordering::SeqCst) {
+        return;
+    }
+
+    let stats = Region::new(GLOBAL);
+    let api_host_names = Settings::api_host_names().join(",");
+    let service_id = Settings::service_id();
+
+    thread::spawn(move || {
+        loop {
+            {
+                let allocator_stats = stats.change();
+                let mem_stats = memory_stats().unwrap_or({
+                    MemoryStats {
+                        physical_mem: 0,
+                        virtual_mem: 0,
+                    }
+                });
+
+                reporter::MEMORY_PHYSICAL_USAGE
+                    .with_label_values(&[&api_host_names, service_id])
```

Comment:
I am not sure if this is the correct label we should apply or not.  Don't change it, but I want to see how the metrics get reported in Grafana to make sure these labels are clearly identifying the metric.